### PR TITLE
Add join hint

### LINKED ISSUE #3688: Feature: Join Hint
### Description

As we entering the optimization stage, for many benchmark queries, we need to fix a join order to isolate the performance bug between computation layer and storage layer. Also, in case a user suffers from a bad join order, we can always fix by giving a fixed join order without changing our cost model or cardinality estimator.

I propose we follow the Postgres way of join hint, e.g.
```
MATCH (a)-[e]->(b)
HINT a JOIN (e JOIN b)
```

### PR DESCRIPTION
**Pull request by andyfengHKU**
# Description

Fixes #3688 

# Contributor agreement

- [x] I have read and agree to the [Contributor Agreement](https://github.com/kuzudb/kuzu/blob/master/CLA.md).

### CODE REVIEW

File: src/include/planner/join_order/join_tree_constructor.h [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        const binder::expression_vector& predicates);
+
+private:
+    const binder::QueryGraph& queryGraph;
+    const PropertyExprCollection& propertyCollection;
+    binder::expression_vector queryGraphPredicates;
+};
+
+} // namespace planner
+} // namespace kuzu
```
Comment (acquamarin) [inline src/include/planner/join_order/join_tree_constructor.h:42]: newline here.

File: src/include/planner/planner.h [hunk 2]
```diff
@@ -26,6 +27,37 @@ namespace planner {
...
+
+struct QueryGraphPlanningInfo {
+    // Predicate info.
+    binder::expression_vector predicates;
+    // Subquery info.
+    SubqueryType subqueryType = SubqueryType::NONE;
+    binder::expression_vector corrExprs;
+    cardinality_t corrExprsCard = 0;
+    // Join hint info.
+    std::shared_ptr<binder::BoundJoinHintNode> hint = nullptr;
```
Comment (acquamarin) [inline src/include/planner/planner.h:44]: Can we use unique pointer there?

File: src/planner/operator/logical_plan_util.cpp [hunk 3]
```diff
@@ -63,33 +78,46 @@ void LogicalPlanUtil::encodeCrossProduct(LogicalOperator* /*logicalOperator*/,
...
+        encodeString += "IndexScan";
+    } else {
+        encodeString += "S";
+    }
+    encodeString +=
+        "(" + scan.getNodeID()->constCast<PropertyExpression>().getRawVariableName() + ")";
+}
+
+void LogicalPlanUtil::encodeFilter(LogicalOperator* op, std::string& encodedString) {
+    //    auto& f = op->constCast<LogicalFilter>();
```
Comment (acquamarin) [inline src/planner/operator/logical_plan_util.cpp:119]: remove

**Review by acquamarin — APPROVED** [review]

**Review by ray6080 — APPROVED** [review]

File: src/include/planner/planner.h (hunk 2 above)
Comment (andyfengHKU) [reply src/include/planner/planner.h:44]: Yes for this PR but not in the soon future where we will need shared_ptr in DP table.

File: src/include/binder/query/reading_clause/bound_match_clause.h [hunk 4]
```diff
@@ -6,20 +6,43 @@
...
           matchClauseType{matchClauseType} {}
 
     QueryGraphCollection* getQueryGraphCollectionUnsafe() { return &collection; }
     const QueryGraphCollection* getQueryGraphCollection() const { return &collection; }
 
     common::MatchClauseType getMatchClauseType() const { return matchClauseType; }
 
+    void setHint(std::shared_ptr<BoundJoinHintNode> root) { hintRoot = std::move(root); }
+    bool hasHint() const { return hintRoot != nullptr; }
+    std::shared_ptr<BoundJoinHintNode> getHint() const { return hintRoot; }
```
[the remark this answers is not in the document]
Comment (andyfengHKU) [reply src/include/binder/query/reading_clause/bound_match_clause.h:40]: See the second comment from @acquamarin

File: src/binder/bind/read/bind_match.cpp [hunk 5]
```diff
@@ -17,10 +40,37 @@ std::unique_ptr<BoundReadingClause> Binder::bindMatchClause(const ReadingClause&
...
+        validateHintCompleteness(*hint, *boundMatch->getQueryGraphCollection()->getQueryGraph(0));
+        boundMatch->setHint(std::move(hint));
+    }
     boundMatch->setPredicate(boundGraphPattern.where);
     return boundMatch;
 }
 
+std::shared_ptr<BoundJoinHintNode> Binder::bindJoinHint(const parser::JoinHintNode& joinHintNode) {
+    if (joinHintNode.isLeaf()) {
+        std::shared_ptr<Expression> pattern = nullptr;
```
[the remark this answers is not in the document]
Comment (andyfengHKU) [reply src/binder/bind/read/bind_match.cpp:57]: As we discussed I have reverted this naming. But fully revert will take a bit of time.

**Pull request merged**